### DISCUSSION TITLE
SNMP connection setup help needed

### DISCUSSION THREAD

**Discussion by Grumpy-Old-Engineer** [nutdotnet/WinNUT-Client#190] [Q&A] [answered]
Synology effectively blew me off with respect to updating their NUT driver(s).  See https://github.com/nutdotnet/WinNUT-Client/discussions/189.  I decided to bypass the NAS and try to talk directly to the UPS via SNMP to see if I can fix the erroneous data display.  I bought a CyberPower RMCard205 for my CP1500PFCLCDa and everything appears to work when I view UPS info in a browser, but I can't get WinNUT to talk to the UPS.  

I keep getting this in the log file:
> 10/4/2024 8:35:03 PM [18188, WinNUT]: Disconnected from Nut Host
> 10/4/2024 8:35:03 PM [18188, WinNUT]: Client UPS_Connect subroutine beginning.
> 10/4/2024 8:35:03 PM [18188, Nut_Socket]: Attempting TCP socket connection to 192.168.3.16:161...

and I get this error message at the bottom of the WinNut client window:
> [Id 41: 10/4/2024 8:35:05 PM] Connection to Nut Host 192.168.3.16:161 Failed: No connection could be made because the target machine actively refused it 192.168.3.16:161

I can see a lot of OIDs when I run FrameFlow SNMP browser, so SNMP on the UPS appears to be working. 
I've tried ports 161 and 3493.  
I've tried both SNMPv1 and SNMPv3 in the UPS.  

Does anyone have any idea what I'm doing wrong?

**Comment by gbakeman** [comment]
Hi again @Grumpy-Old-Engineer! ☺️

Sorry again about the difficulties with Synology. Someone else here was using SNMP (and the same management card as well) in #126 in case that's of any use to you. I spent a few minutes looking into this further, and while that management card does support many protocols, critically it does not support the NUT protocol which is what WinNUT relies on to talk to a NUT server. I honestly would've expected a different error from WinNUT, although the idea's the same. Conveniently however, the NUT server does appear to have an SNMP driver so you can connect your NUT server to the UPS that way, then hook WinNUT back up to the NUT server. My search query was _RMCard205 nut_ which returned several discussion threads of people getting their NUT server to talk to the UPS.

I feel bad about all of the effort you're going through to fix this bug... are you sure something small like a Pi wouldn't be sufficient for running a vanilla build of NUT and monitoring your UPS by USB?

**Reply by Grumpy-Old-Engineer** [reply]
Thanks for the reference, but.... I understand that _if_ I can get SNMP to work I might still have data content problems.  My issue here is that I can't get SNMP to connect to the UPS at all, so I have no data to evaluate.  The referenced issue reads like the OP was using SNMP with the RMCard205, not USB, so he was able to connect to the UPS.  Web searches have turned up issues in other forums where people were having data content issues, but they had to be connected to have those issues.  I'm hoping someone out there has the "duh, you need to do this to connect via SNMP" answer.

As for an RPi solution, that would have to be procured and built.  I am trying to use things that are slightly more readily available - RMCard on eBay and updated with the latest f/w (1.4.1) was an easy first try - right up to the point that I couldn't get the SNMP connected.

**Reply by gbakeman** [reply]
Just to be clear; it sounds like the SNMP server is working fine and you can see that data is being read correctly from the UPS (inside a webserver on the SNMP card its self?), but cannot get WinNUT to talk to the SNMP card at all?

To put what I meant earlier another way, WinNUT only speaks NUT, so there needs to be a NUT server inbetween WinNUT and the UPS (or SNMP card, as the case may be.) Unfortunately, it looks like the SNMP card does not run its own NUT server so that task still has be outsourced to somewhere else. It may be possible for you to configure the Synology NUT server to connect using its SNMP driver rather the Cyberpower USB driver that seems to be the culprit from before.

Sorry if I'm just misunderstanding here.

**Comment by Grumpy-Old-Engineer** [comment]
> Just to be clear; it sounds like the SNMP server is working fine and you can see that data is being read correctly from the UPS (inside a web server on the SNMP card its self?), but cannot get WinNUT to talk to the SNMP card at all?

Yes, this is correct, I can access the UPS using a web browser.  My SNMP browser shows me 486 OIDs from the UPS, so I believe that the RMCard's SNMP functionality is working.

> To put what I meant earlier another way, WinNUT only speaks NUT, so there needs to be a NUT server in between WinNUT and the UPS (or SNMP card, as the case may be.) 

OK, this means I missed the boat.  I was under the impression that WinNut was using SNMP, and could connect directly to any UPS that supported SNMP.  

> It may be possible for you to configure the Synology NUT server to connect using its SNMP driver rather the Cyberpower USB driver that seems to be the culprit from before.

Chuckle chuckle.... I actually tried this before I got an RMCard.  Whatever Synology is doing to translate the UPS/SNMP data and deliver it via their Nut server provides even worse results than their UPS/USB implementation.  When I configured the NAS to access the UPS using SNMP, the only non-zero values on the WinNut Client display were Frequency and Battery Voltage.  UPS/USB _and_ UPS/SNMP, Synology's performance is severely lacking. 

@gbakeman - I greatly appreciate the advice you've provided and the patience you've shown.  Based on all of your comments, both here and in https://github.com/nutdotnet/WinNUT-Client/discussions/189, it looks like I (still) either deal with displays that are crippled by Synology's lackluster performance, or I figure out how to build my own Nut server and pick somewhere to run it.

**Reply by gbakeman** [reply]
Yikes... I didn't realize that Syno's NUT server was that bad! Ok, I see your predicament now. Sorry again that this has been so difficult.

**Comment by Grumpy-Old-Engineer** [comment, marked as answer]
While sometimes it's enjoyable to gripe about other/additional problems, I started this discussion "just" asking for help to get Nut to talk SNMP to a CyberPower UPS that's using a CyberPower RMCard205 card.  It seems like everybody that has provided any kind of tutorial was using USB, not SNMP, but because of Synology's antiquated Nut in DSM I wanted to find another way to do things.

After spending too much time web searching and getting nowhere I went back to basics - RTFM.  After several dozen iterations of [read this, try that, go back and RTFM again] I painfully realized that there doesn't appear to be any room for not-quite-right in the UPS definitions in ups.conf.  Until I got to this point nothing worked for me:
        [UPS1]
                driver = snmp-ups
                port = 192.168.3.15
                community = private
                snmp_version = v1
                desc = "CyberPower CP1500PFCLCD"

This looks simple now, but hindsight is usually 20/20.
- The UPS name doesn't matter (stay away from spaces and special characters).
- The IP address in the port parameter doesn't matter as long as it's the IP address that the UPS is configured to use.  I will admit that I quit when things started to work - I didn't go back and try using a DNS name.
- The community parameter is needed.
- The community value doesn't matter as long as it is **_exactly_** what the UPS is configured to look for.  For my UPS case matters (grumble grumble).
- The snmp_version should default to SNMPv1, and an entry is not supposed to be needed, but things didn't work without it.  I'm not sure where I was in my read/try cycles when things started to work, but "snmp_version = v1" was in there at the time and I quit messing with it while I was ahead (or at least no longer falling behind as fast).  I did not try to figure out SNMPv3, and since this is all internal to my home I don't really need it.  
- The description value doesn't matter, but I'll admit I never tried without including something
- DO NOT include the vendorid parameter as needed for the same UPS when connecting via USB

I'm finally running well enough to start contemplating how I want to set up all of my upsmon.conf settings.  Hopefully that will go smoother.

**Reply by gbakeman** [reply]
Well done and thanks again for your all of your effort here. Are you still using the Synology NUT server (with the SNMP driver) or did you get another NUT server spun up?

**Comment by Grumpy-Old-Engineer** [comment]
The short answer – I have decided to go with 2 servers.
1. I’m going to keep using DSM’s vintage semi-custom implementation of Nut with USB UPS, but just for the NAS itself.   DSM’s SNMP UPS implementation is worse than their USB UPS implementation, I know that the USB UPS implementation works to shut down the NAS as needed, and I just don’t want to risk the NAS with something I don’t really trust.  
2. I’m going to run Nut on an Ubuntu Server in a VM on the NAS, monitor the same UPS (and others) via SNMP, and manage everything but the NAS from that instantiation of Nut.

Here’s some additional info for anyone that finds this in the future while searching for solutions to similar problems.  Good luck!

As for using DSM (both versions 6.2.2 and 7.2.2 gave me the same problems) with Nut/SNMP, here’s a list of seemingly random pieces of info:
- If an UPS is physically connected via USB to the NAS, the DSM UPS function will only use the USB UPS option.
- In order to try to run DSM UPS with an SNMP UPS, the USB UPS must be physically disconnected.
- In order for Nut in a VM running on the NAS to access a USB UPS, all of the DSM’s UPS support must be disabled in /etc.defaults/synoinfo.conf (reboot required).  When disabled, DSM will not attempt to access any UPS – USB, SNMP, or another Nut server (labeled Synology UPS server on the DSM control panel).
- DSM UPS serves data from a USB UPS to WinNut on a Windows PC, but Output Voltage value is wrong, and the values for Manufacturer and Firmware need work.
![image](https://github.com/user-attachments/assets/d64bc013-8010-4319-b03d-e15fa2509a50)
- DSM UPS serves data from an SNMP UPS to WinNut on a Windows PC with different (less useful) results, but at least the Manufacturer and Firmware values are good.
![image](https://github.com/user-attachments/assets/338b214f-1524-4203-a1ff-990062afd1eb)
- DSM UPS will access a Nut Server running on Ubuntu Server elsewhere on the network, BUT the associated instance of ups.conf must have an UPS named “UPS”.  I have not been able to determine how the DSM UPS function gets this value, so I have no clue how to change it.  In this mode the DSM UPS will not serve data to WinNut running on a PC (and probably not anywhere else).

Based on the DSM UPS server performance when trying to use SNMP UPS, I decided to try running Nut on something else.  My first attempt was Ubuntu in a VM running on the NAS.  doo…. my NAS was a bit underpowered for that so I switched to Ubuntu Server.  Nut ran acceptably in the VM, but the need to edit /etc.defaults/synoinfo.conf and reboot the NAS every time I wanted to test with or without a USB connection got to be a real PIA.  Subsequently I built an Ubuntu Server on a Dell OptiPlex and ran Nut there also.  This was my test setup:
![House_UPS_test_config](https://github.com/user-attachments/assets/63a7e713-9200-4fbf-9ccb-57330f597a15)
Now, don’t everyone come out of the woodwork and slam my configuration…. I only put this together to try to figure out how to make DSM UPS, Nut, and WinNut work and play (relatively) well together.  In the above configuration:
- DSM UPS is monitoring UPS1 via USB
- Nut Server in the VM is monitoring both UPS1 and UPS2 via SNMP
- Nut Server in the OptiPlex is monitoring both UPS1 and UPS2 via SNMP
- Nut Server in the OptiPlex is monitoring UPS3 via USB
- DSM UPS provides UPS1 data (with errors) to WinNut
- Nut Server in the VM provides UPS1 and UPS2 data to WinNut
- Nut Server in the OptiPlex provides UPS1, UPS2, and UPS3 data to WinNut

Even the browser GUI works:
![image](https://github.com/user-attachments/assets/b553616d-1ecf-4379-abf3-8f4b11e881aa)

The next step will be to decide what is wanted and/or needed with respect to notification and command actions, and figure out how to make those work, but that’s out of scope for this discussion thread…..

**Reply by gbakeman** [reply]
Once again, incredible work on researching and writing this up. Your configuration is becoming close to what mine is now.

I have a generic computer of some kind that's running most all of my services. My Synology is running a minimal amount of services, and is only intended as a backup destination for my data. My router now is a small Dell box just like what you have, running OPNsense. Since that device is the only one (as far as I know) designed to power off at a moment's notice, it's also what I have running a NUT server to monitor the UPS powering these three devices. I'm realizing more and more that Synology's software is not great or designed for a very specific purpose, so I'm trying to limit the involvement of their NUT server/client as much as possible. I currently have it connected as a client to the NUT server on the router, but even then I'm running into a bug with the Synology where it will place the NAS into some kind of suspended state where it's able to lose power. The problem with that is it keeps the drives spinning (consuming more power from my UPS), and if the power returns, there's no way to reboot the NAS besides doing a physical power off of the device. My ultimate solution will involve removing the Synology software from as much control as possible. I'll likely just have a script that logs into the SSH server on the NAS and run a poweroff command.

I hope you make another discussion when you've decided on what to do with your setup!
